compiler/AST/iterator.cpp

```diff
@@ -112,43 +112,81 @@ isSingleLoopIterator(FnSymbol* fn, Vec<BaseAST*>& asts) {
     return NULL;
   BlockStmt* singleFor = NULL;
   CallExpr* singleYield = NULL;
-  forv_Vec(BaseAST, ast, asts) {
+  forv_Vec(BaseAST, ast, asts)
+  {
+    // If a yield statement,
     if (CallExpr* call = toCallExpr(ast)) {
       if (call->isPrimitive(PRIM_YIELD)) {
         if (singleYield) {
+          // We already saw a yield stmt.  This is the second one, so fail.
           return NULL;
-        } else if (BlockStmt* block = toBlockStmt(call->parentExpr)) {
-          // NOAKES 2014/11/18 It is strange that this ignores DOWHILE
-          if (block->isForLoop()     == true ||
-              block->isCForLoop()    == true ||
-              block->isWhileDoStmt() == true) {
-            singleYield = call;
-          } else {
+        }
+        
+        // Select yield statements whose parent expression is a loop statement
+        // (except for dowhile statements, for some reason....
+
+        // This test is not logically related to the preceding quick-exit, so
+        // putting "else" here would be misleading.
+        if (dynamic_cast<LoopStmt*>(call->parentExpr) != NULL) {
```

Comment:
Easier than looking at the undocumented interface or source code of toLoopStmt, to see if it actually does what I want.  Now that I've tested with dynamic_cast (safe and sure), I can substitute toLoopStmt, and see if anything breaks.